fdbclient/S3Client.actor.cpp

```diff
@@ -166,27 +351,207 @@ ACTOR Future<Void> copyUpBulkDumpFileSet(std::string s3url,
+			    .detail("Attempt", attempts);
+
+			wait(delay(retryDelayMs / 1000.0));
+		}
+	}
+	throw operation_failed();
+}
+
+ACTOR static Future<Optional<std::string>> getObjectRangeMD5(Reference<S3BlobStoreEndpoint> endpoint,
+                                                             std::string bucket,
+                                                             std::string objectName,
+                                                             int64_t offset,
+                                                             int64_t size) {
+	try {
+		HTTP::Headers headers;
+		headers["Range"] = format("bytes=%lld-%lld", offset, offset + size - 1);
+
+		Reference<HTTP::IncomingResponse> response = wait(endpoint->doRequest("HEAD",
+		                                                                      objectName, // resource
+		                                                                      headers, // headers
+		                                                                      nullptr, // content
+		                                                                      0, // content length
+		                                                                      { 200 })); // expected response codes
+
+		auto it = response->data.headers.find("Content-MD5");
+		if (it != response->data.headers.end()) {
+			return Optional<std::string>(it->second);
+		}
+
+		return Optional<std::string>();
+	} catch (Error& e) {
+		// If we can't get the MD5, just return empty optional
+		TraceEvent(SevWarn, "S3ClientGetMD5Failed")
+		    .error(e)
+		    .detail("Bucket", bucket)
+		    .detail("Object", objectName)
+		    .detail("Offset", offset)
+		    .detail("Size", size);
+		return Optional<std::string>();
+	}
+}
+
 // Copy down file from s3 to filepath.
 ACTOR static Future<Void> copyDownFile(Reference<S3BlobStoreEndpoint> endpoint,
                                        std::string bucket,
-                                       std::string resource,
-                                       std::string filepath) {
+                                       std::string objectName,
+                                       std::string filepath,
+                                       PartConfig config = PartConfig()) {
 	TraceEvent("S3ClientCopyDownFileStart")
 	    .detail("filepath", filepath)
 	    .detail("bucket", bucket)
-	    .detail("resource", resource);
-	std::string content = wait(endpoint->readEntireFile(bucket, resource));
-	auto parent = std::filesystem::path(filepath).parent_path();
-	if (parent != "" && !std::filesystem::exists(parent)) {
-		std::filesystem::create_directories(parent);
+	    .detail("objectName", objectName);
+	state int64_t fileSize = wait(endpoint->objectSize(bucket, objectName));
+	state Reference<IAsyncFile> file;
+	state std::vector<PartState> parts;
+
+	try {
+		// Create or truncate the output file. Create subdirs if needed.
+		platform::createDirectory(parentDirectory(filepath));
```

Comment:
Is it possible that the parentDirectory(filepath) exists and the folder contains some outdated file?

Reply:
Dunno. Paths are all hashed to avoid any such clash. If the file exists, we'll overwrite it. If the file is not mentioned in the manifest, it'll be ignored. If it is a manifest file, will we read it?